lsql/judge/models.py

```diff
@@ -414,3 +428,102 @@ class Submission(models.Model):
 
     def __str__(self):
         return f"{self.pk} - {self.user.email} - {self.veredict_code}"
+
+
+class AchievementDefinition(models.Model):
+    """Abstract class for Achievements"""
+    name = models.TextField(max_length=5000, validators=[MinLengthValidator(1)], blank=True)
+    description = models.TextField(max_length=5000, validators=[MinLengthValidator(1)], blank=True)
+
+    def check_and_save(self, user):
+        """Raise a NotImplementedError, declared function for its children"""
+        raise NotImplementedError
+
+    def check_user(self, usr):
+        """Check if an user have the achievement"""
+        achievements_of_user = ObtainedAchievement.objects.filter(user=usr, achievement_definition=self).count()
+        return achievements_of_user > 0
+
+    def refresh(self):
+        """Delete the achievement and check if any user have it"""
+        ObtainedAchievement.objects.filter(achievement_definition=self).delete()
+        all_users = get_user_model().objects.all()
+        for usr in all_users:
+            self.check_and_save(usr)
+
+    def __str__(self):
+        """String for show the achievement name"""
+        return f"{self.name}"
+
+
+class ObtainedAchievement(models.Model):
+    """Store info about an obtained achievement"""
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
+    obtained_date = models.DateTimeField()
+    achievement_definition = models.ForeignKey(AchievementDefinition, on_delete=models.CASCADE)
+
+
+class NumSolvedAchievementDefinition(AchievementDefinition, models.Model):
+    """Achievement by solving a number of problems"""
+    num_problems = models.PositiveIntegerField(default=1, null=False)
+
+    def check_and_save(self, user):
+        """Check if an user is deserving for get an achievement, if it is, save that"""
+        if not self.check_user(user):
+            corrects = Submission.objects.filter(veredict_code=VeredictCode.AC, user=user).distinct("problem").count()
+            if corrects >= self.num_problems:
+                order_problem_creation_date = Submission.objects.filter(creation_date__in=Subquery(
```

Comment:
¿Puedes añadir un breve comentario explicando esta consulta compleja?

Reply:
Bueno, y también las demás que sean similares

Reply:
Para no repetir más, ¿este comentario está bien o tengo que poner algo más?

_First submission of each Problem that user have VeredictCode.AC. Ordered by 'creation_date'_

Reply:
Me parece buen comentario, y si puedes dar alguna pincelada del Subquery perfecto